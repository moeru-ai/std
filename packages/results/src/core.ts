export interface Err<E> {
  __type__: 'err'
  error: E
}

export interface None {
  __type__: 'none'
}

export interface Ok<T> {
  __type__: 'ok'
  value: T
}

export type Option<T> = None | Some<T>

export type Result<T, E> = Err<E> | Ok<T>

export interface Some<T> {
  __type__: 'some'
  value: T
}

export const err = <T, E>(error: E): Result<T, E> => ({
  __type__: 'err',
  error,
})

export const none: Option<never> = {
  __type__: 'none',
}

export const ok = <T, E>(value: T): Result<T, E> => ({
  __type__: 'ok',
  value,
})

export const some = <T>(value: T): Option<T> => ({
  __type__: 'some',
  value,
})
