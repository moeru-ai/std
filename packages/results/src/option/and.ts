import type { Option } from '../core'

import { isSome } from './is'

export const and = <T1, T2>(o: Option<T1>, res: Option<T2>): Option<T2> =>
  isSome(o)
    ? res
    : o

export const andThen = <T1, T2>(o: Option<T1>, onSome: (s: T1) => Option<T2>): Option<T2> =>
  isSome(o)
    ? onSome(o.value)
    : o

// eslint-disable-next-line sonarjs/no-identical-functions
export const andThenAsync = async <T1, T2>(o: Option<T1>, onSome: (s: T1) => Promise<Option<T2>>): Promise<Option<T2>> =>
  isSome(o)
    ? onSome(o.value)
    : o
