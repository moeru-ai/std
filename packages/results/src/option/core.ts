export interface None {
  __type__: 'none'
}

export type Option<T> = None | Some<T>

export interface Some<T> {
  __type__: 'some'
  value: T
}

export const none: Option<never> = {
  __type__: 'none',
}

export const some = <T>(value: T): Option<T> => ({
  __type__: 'some',
  value,
})
