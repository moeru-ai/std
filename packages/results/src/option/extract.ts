import type { Option } from './core'

import { isSome } from './is'

/** @experimental */
export const extract = <T>(o: Option<T>): T | undefined =>
  isSome(o)
    ? o.value
    : undefined
