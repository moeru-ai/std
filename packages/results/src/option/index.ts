export * from './and'
export * from '../core'
export * from './extract'
export * from './is'
export * from './map'
export * from './match'
export * from './ok'
export * from './or'
export * from './unwrap'
export * from './wrap'
