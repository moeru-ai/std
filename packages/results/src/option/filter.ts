import type { Option } from '../core'

import { none } from '../core'
import { isSome } from './is'

export const filter = <T>(o: Option<T>, check: (v: T) => boolean): Option<T> =>
  isSome(o) && check(o.value)
    ? o
    : none

export const filterAsync = async <T>(o: Option<T>, check: (v: T) => Promise<boolean>): Promise<Option<T>> =>
  isSome(o) && await check(o.value)
    ? o
    : none
