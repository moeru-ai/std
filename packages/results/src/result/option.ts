import type { Option, Result } from '../core'

import { none, some } from '../core'
import { isErr, isOk } from './is'

// Originally Result.ok(), renamed optionOk due to naming conflict
export const optionOk = <T, E>(r: Result<T, E>): Option<T> =>
  isOk(r)
    ? some(r.value)
    : none

// Originally Result.err(), renamed optionErr due to naming conflict
export const optionErr = <T, E>(r: Result<T, E>): Option<E> =>
  isErr(r)
    ? some(r.error)
    : none
