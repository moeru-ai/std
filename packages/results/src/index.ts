import type { None, Option, Some } from './option'
import type { Err, Ok, Result } from './result'

import * as option from './option'
import * as result from './result'

export {
  Err,
  None,
  Ok,
  Option,
  option,
  result,
  Result,
  Some,
}
