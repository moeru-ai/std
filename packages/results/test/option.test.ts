import { describe, expect, it } from 'vitest'

import type { Option } from '../src'

import { none, some } from '../src'
import { match, unwrap } from '../src/option'

describe('@moeru/results/option', () => {
  it('basic', () => {
    const checkedDivision = (dividend: number, divisor: number): Option<number> =>
      divisor === 0
        ? none
        : some(dividend / divisor)

    const tryDivision = (dividend: number, divisor: number) =>
      match(
        checkedDivision(dividend, divisor),
        quotient => `${dividend} / ${divisor} = ${quotient}`,
        () => { throw new Error(`${dividend} / ${divisor} failed!`) },
      )

    expect(tryDivision(4, 2)).toBe('4 / 2 = 2')
    expect(() => tryDivision(1, 0)).toThrowErrorMatchingSnapshot()

    const optionalFloat = some(0)
    expect(unwrap(optionalFloat)).toBe(0)
    expect(() => unwrap(none)).toThrowErrorMatchingSnapshot()
  })
})
