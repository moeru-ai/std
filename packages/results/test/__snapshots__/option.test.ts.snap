// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@moeru/results/option > basic 1`] = `[Error: 1 / 0 failed!]`;

exports[`@moeru/results/option > basic 2`] = `[Error: called \`option.unwrap()\` on a \`None\` value]`;
